[RFC 0051] Mark stale nixpkgs issues

ryantm opened the pull request:
Summary:

> Mark stale Nixpkgs issues and pull requests on GitHub using an application provided by GitHub.

[Rendered](https://github.com/ryantm/rfcs/blob/stale/rfcs/0051-mark-stale-issues.md)

timokau commented on rfcs/0051-close-stale-issues.md:
```diff
@@ -0,0 +1,74 @@
+   markComment: >
+     This issue has been automatically marked as stale because it has not had
+     recent activity. It will be closed if no further activity occurs. Thank you
+     for your contributions.
```
I think the framing here is very important to (1) don't make anyone feel like they're just a drag on the project, make clear that it's totally fine to just keep bumping an issue; (2) give constructive feedback on how to get attention for an issue, like actively searching for maintainers / previous people that touched the code and pinging them.

michaelpj replied:
I think this is a great point. Phrasing it as helping the user seems much more likely to turn out well. "It looks like this issue hasn't attracted much attention. Here are n suggestions for how to progress this issue further."

zimbatm commented on rfcs/0051-close-stale-issues.md:
```diff
@@ -0,0 +1,74 @@
+[motivation]: #motivation
+
+The Nixpkgs GitHub page has a large number of open issues causing
+community angst and misrepresenting the responsiveness of the project.
```
It would be good to find metrics that better represent our project's health and communicate that instead. It seems quite natural to me that due to the size of nixpkgs we have many known problems that needs to be addressed.

gloaming replied:
Yes, when I first came here, I remember being shocked by the numbers. But now they just seem realistic. We'd all like them to go down, but I feel like closing issues prematurely would cause more angst, not less. Imagine how it would feel when, in addition to having your issue "ignored", you then suffer the further indignity of having it closed by a bot!

It's already easy to get the impression that nobody cares about your issue or PR. Wouldn't closing outstanding issues would basically confirm that?

zimbatm replied:
I think it really depends on the bot's abilities. For example I am never annoyed by the existence of ofborg as it helps me move my PR along. Even when it was posting build results as comments, I was happy to get the notifications because it was an asynchronous prompt for me to react on the feedback.

Stale bots are less clear cut. It depends how precise their heuristics are. If implemented correctly I can see how it could prompt people to react and move the issue or PR along. But if they generate too much noise it can also lead to notification fatigue. It really depends on the implementation.

gloaming replied:
Sure, I wouldn't be opposed to a kind of ping-bot. But I understood this RFC as being primarily about closing issues, with the bot being an implementation detail.

I actually wasn't thinking about the case where the _submitter_ has not responded; in that case, I'm not opposed to automated closing if it's a "support ticket" kind of issue where others can't reproduce. That would help make the list of open issues more relevant, I think.

timokau replied:
Maybe it would be an option to only ask for a follow up if the last message is not from the submitter? This way the submitter would only have to say "no, this is still an issue" once and the bot won't bother them again until someone else asks for some followup.

Ekleog replied:
Either this, or maybe it would be possible to have the bot only trigger on `9.needs: reporter feedback`?

That said, my experience using codetriage is that the issues I click on usually would not fall under this label -- it's only maybe one in four or five. But it does happen that when I ping from triage to re-request the requested information, the reporter just states they've stopped using NixOS altogether, and then it's possible to close the ticket.

All that to say: I think a stale bot would make sense, but more for the “regularly poke people with a message to help them figure out how to push it forward” part of it and less for the “automatically close issues” part of it. If it's done this way, we should be careful to not have it poke too many issues on the first start, or maintainers would be overwhelmed by people directly reaching to them at once, though.

zimbatm replied:
Another scenario is when a PR gets stuck because the author and the reviewers don't agree on the outcome. In that case, if the bot is pinging, it's not bringing in new information that could help move the PR forward. If it's closing it favours the status quo over innovation.

Melkor333 replied:
Not necessarily. If for example the bot links to a page containing as many suggestions as possible on how to move on in such situations (which in my opinion should be a part of this RFC), I would consider it very helpful.

michaelpj commented on rfcs/0051-close-stale-issues.md:
````diff
@@ -0,0 +1,74 @@
+
+   ```
+   # Number of days of inactivity before an issue becomes stale
+   daysUntilStale: 60
````
I think this should be *much* higher, perhaps ~1 stable release of NixOS? There are a lot of useful issues older than 60 days!

blaggacao replied:
_stale_ does not mean "not-useful" &mdash; why do you assume it?

Profpatsch commented:
I am completely against this idea. In my experience, this this the archetypal way of trying to solve social issues with technical solutions.

Apart from “making that number go down and feel better”, introducing arbitrary time-outs for mostly still valid issues is a sure way to lose potential contributors.
In addition, there is immense value in having a database of known and reported problems, which this RFC is going to just waste.

Github provides good tooling to filter issues down to the ones that are interesting to review, no need to confuse “issue open” with “failure of project”.

michaelpj commented on rfcs/0051-close-stale-issues.md:
```diff
@@ -0,0 +1,74 @@
+   # Number of days of inactivity before an issue becomes stale
+   daysUntilStale: 60
+   # Number of days of inactivity before a stale issue is closed
+   daysUntilClose: 7
```
Likewise here, maybe a month?

suhr replied:
I would prefer *never*. Let humans to decide.

adisbladis commented:
To add to what @Profpatsch is saying my previous issues that has gone "stale" in other repos and it's been a feeling of incredible frustration to have them auto-closed just to artificially decrease an issue number.
The issue is still there whether you close it or not.

Another thing I have noticed about Nixpkgs issues is that some of them are very long-running for technical reasons.

One example of this is https://github.com/NixOS/nixpkgs/issues/33798. Just because there isn't much happening on this issue doesn't mean that it's an issue that we want to close, it's still an ambition to drop Qt 5.6 at _some point_, but as of right now a few packages requires it to run so we don't drop it.

I'd be angry and frustrated if such issues were auto-closed.

jollheef commented on rfcs/0051-close-stale-issues.md:
````diff
@@ -0,0 +1,74 @@
+
+   ```
+   # Number of days of inactivity before an issue becomes stale
+   daysUntilStale: 60
````
I think it must be not less than 6 months. One year is even better.

In this way, it'll help get rid of issues that can't be reproduced anymore, but also it'll not cause frustration of issue creator.

suhr commented:
The idea is to ping old issues and close one that do not respond. I'm not sure if automatic closing is a good idea, but notification and labeling is, if it's done rarely (once in a year).

Infinisil commented:
Yeah I'm against this as well. Just recently @peti closed [an issue](https://github.com/NixOS/nixpkgs/issues/32005#issuecomment-496315702) on me, seemingly because of the same reason this RFC tries to argue for, even though it's still totally valid, which also heavily demotivates me to work on it. I don't want this to become the norm.

peti commented:
> [Closing the issue] heavily demotivated me to work on it. I don't want this to become the norm.

I honestly don't want to sound dismissive of your point of view. There is just one thing that I don't understand. If you have been motivated before to work on that issue, then why had there been no progress and no activity whatsoever for 15+ months at the time when I just closed it? I mean, when you say you feel demotivated working on it *now*, I'm not sure how to parse that. What was your motivation level before I closed it exactly?

IMHO, if an issue has been completely inactive for, say 12 months, then it's dead and I see no point in keeping it open. I would tag it with OBSOLETE or AUTO-CLOSED or whatever and close it.

If you actually care about an issue, then surely you should be able to make some kind of headway within a freaking year, no?

suhr commented:
The purpose of issues is to keep track of existing problems. Closing issues for existing problems defeats this purpose as well as keeping track of issues that nobody can reproduce anymore.

globin commented:
I nominate @grahamc and myself for shepherding

peti commented:
> The purpose of issues is to keep track of existing problems.

Not everyone agrees that this is their purpose.

emilazy commented:
my 2¢: I really hate it when projects do this and I don't think it improves software quality at all.

Ekleog commented:
@peti 
> why had there been no progress and no activity whatsoever for 15+ months at the time when I closed it?

The thing is, issues are not only meant for *developers*. They are meant for *users* to report issues so that developers can fix them. And for a regular user, it is very normal that in 15+ months they make no progress -- they would maybe need to learn everything from scratch just to make headway on this. (and, in this specific case, I guess the priority was just below other more-important stuff)

Saying that we should close issues when, after a bot ping, no one answers for some time kind of makes sense (no one is interested in it any longer, and it quite possibly doesn't exist any longer). Saying that we should close issues when there is no progress made means that we decide to have no way for non-developers to report issues.

>> The purpose of issues is to keep track of existing problems.
>
> Not everyone agrees that this is their purpose.

I think it is the first time I hear of someone opposing this statement. What is their purpose, for you?

gloaming commented:
> The purpose of issues is to keep track of existing problems.

> Not everyone agrees that this is their purpose.

> The thing is, issues are not only meant for developers. They are meant for users to report issues so that developers can fix them.

I think this argument arises in part from a missing (conceptual) distinction between _issues_ and _tickets_. Consider that a single underlying issue may manifest for multiple users in ways that cause them to file multiple reports.

Currently, the Nixpkgs GitHub is a main venue for user support, so we have both; but that's not a reason to give them the same blanket treatment. (That said, it is also not appropriate to mark unresolved tickets as resolved, and a sophisticated support system distinguishes between resolved and expired.)

GitHub has only a binary open/closed status, so how you think that status should be used is a direct result of what you think that the items in the GitHub tracker _are_.

peti commented:
@Ekleog,

> For a regular user, it is very normal that in 15+ months they make no progress.

I am sorry, but that is not my experience. You say it's *normal* for people to actively work on an open source software issue without any observable progress whatsoever for 15+ months, but I don't believe that. My experience is that once a ticket has gone over a year without any activity, chances that it will be fixed in the future are very, very slim. I don't think it's *normal* at all for tickets to get picked up again after lying there, dead, for over a year. I dare say, it hardly ever happened in my free software career. That's why people write things like the Github stale bot in the first place, so that those dead tickets can be closed.

> Saying that we should close issues when, after a bot ping, no one answers for some time kind of makes sense.

OK, we both agree. I believe that this is the proposal made in this RFC? The bot closes issues who remain inactive of *X* days even after the *ping*, right? That sounds perfectly reasonable.

> Saying that we should close issues when there is no progress made means that we decide to have no way for non-developers to report issues.

Oh, come on, that's is not true. Non-developers are reporting issues in Nixpkgs all the time!!!! Nobody is going to prevent them from doing that.


> > > The purpose of issues is to keep track of existing problems. 
> > 
> > Not everyone agrees that this is their purpose.
> 
> I think it is the first time I hear of someone opposing this statement. What is their purpose, for you?

I perceive Github issues as a way to track important, actionable tasks that someone is currently working on or that someone *should* be working on in the near future. The issue tracker is a way to coordinate the work done by the developers, so that everyone knows what the others are working on and we avoid doing the same thing multiple times by accident. Tickets that lie around for years without anyone even attempting to solve them don't meet those criteria. Clearly, those issue are not actionable or not important or neither of the two.

emilazy commented:
It's rare that someone will work continuously on an issue for 15 months without visible process, but I think it's hardly uncommon in open source for an issue to lie around for 15 months until someone with the interest, expertise and time decides to investigate and fix it. I think incrementalism is the only way a project the size of nixpkgs can advance.

peti commented:
> I think it's hardly uncommon in open source for an issue to lie around for 15 months until someone with the interest, expertise and time decides to investigate and fix it.

Can you point out any concrete examples in the Nixpkgs issue tracker to back up that opinion?

Zimmi48 commented:
@peti This is an easy challenge. This kind of issues is easy to find, just in the issues that were closed in the last month we can find some examples:
- https://github.com/NixOS/nixpkgs/issues/26776: opened for more than two years, no activity for almost a full year, before it was fixed
- https://github.com/NixOS/nixpkgs/issues/18720: no activity for almost three years before a PR fixed "most of these points"
- https://github.com/NixOS/nixpkgs/issues/25025: no activity for over two years before someone ask for an update and this gets fixed
- https://github.com/NixOS/nixpkgs/issues/29095: created almost two years ago, no activity for almost a year before it gets fixed
- https://github.com/NixOS/nixpkgs/issues/33282: no activity for over a year before someone says they are also interested and this gets fixed

Ekleog commented:
@peti 
> You say it's normal for people to actively work on an open source software issue without any observable progress whatsoever for 15+ months, but I don't believe that.

It's a good thing I did *not* say that, then. To quote myself, “for a regular user [as opposed to developer], it is very normal that in 15+ months they make no progress -- they would maybe need to learn everything from scratch just to make headway on this”

The regular user is not going to actively work on an open-source software issue for 15+ months. They are going to shoot the issue, and let people with relevant knowledge fix it, when such people will come onto the issue.

Being better at pointing people with relevant knowledge to relevant issues is what we need to be better at, and I hope a stale-bot-like thing can do it properly without having this hugely-annoying-for-the-bug-reporter behavior of just closing things no one decides to work on.

peti commented:
> @peti This is an easy challenge. This kind of issues is easy to find, just in the issues that were closed in the last month we can find some examples.

Interesting. Clearly you were right. I hardly ever see cases like those myself, but apparently that does not mean that they don't exist.

zimbatm commented:
Would it make sense to get more data points about issues and PRs?

One danger I see of this RFC is that it's natural to judge the RFC based on our personal history. Maybe if we make it more about data then the discussions can become more specific and objective.

I think we all agree that we have many issues open and that automation is a good thing to leverage (that's why we are developers isn't it? :D). All we seem to disagree on is on the implementation and how it should look like.

Zimmi48 commented:
>Would it make sense to get more data points about issues and PRs?

It would. What would be the questions that you would like to use data to answer?

<!--

>I think we all agree that we have many issues open and that automation is a good thing to leverage (that's why we are developers isn't it? :D). All we seem to disagree on is on the implementation and how it should look like.

What I believe would be useful to have would be some guarantee that the status of an issue is correctly reflected by its labels. In the Coq project (which uses label categories based on the nixpkgs model), we have recently introduced a "needs: triage" label, although this label isn't used much yet.
A bot could be run to automatically post a comment asking whether the current status is correctly reflected by the labels, and add such a "needs: triage" label after a certain delay of activity.
But this would require that participants get some issue triaging privileges, or that the bot can be used to modify the status of an issue via a comment, and update the labels.

-->

Profpatsch commented:
Triaging (e.g. going through stale issues and asking for an update) is the thing that is best to automate. But I’d still hesitate to auto-close, even if some time has passed after a ping.

suhr commented:
> What would be the questions that you would like to use data to answer?

For example, it would be nice to know the distribution of issues lifetimes.

Zimmi48 commented:
@suhr Here you go: https://github.com/Zimmi48/nixpkgs-issues

To summarize, on average 25% of issues are closed within 2 days, 50% are closed within 50 days, and 75% are closed within 500 days.

ryantm commented:
It sounds like most people would be in favor of the triage and marking stale aspect, but there is not as much support for the auto-close aspect. Fortunately, the stale app supports that by setting `daysUntilClose` to `false`. Given the feedback here, I would be in favor of that, combined with a more helpful message when it is marked stale.

asymmetric commented:
What’s the purpose of marking an issue as `stale`?

8573 commented:
> What’s the purpose of marking an issue as *stale*?

I can offer only my own speculation, but I imagine it might help someone who's looking for a way to contribute find issues that no-one is solving.

domenkozar commented:
> The Nixpkgs GitHub page has a large number of open issues causing
community angst and misrepresenting the responsiveness of the project.

I've also seen languages (like Elm) where issues are closed even if they are relevant, which results into angst.

The core of the issue is that number of open issues is a vanity metric, what matters are relative trends week-to-week, etc.

While the current "issue is closed" means it's either fixed or won't fix, with this change the difference between open and closed becomes less useful.

I'm -1 for introducing this, instead we should come up with better metrics and point anxious people there to see that everything is OK.

edolstra commented:
@Profpatsch @peti Since you have strong feelings about this, would you be interested in joining the shepherd team for this PR/

peti commented:
> What’s the purpose of marking an issue as _stale_?

I'm by no means a Github expert, but I suppose that label would allow you to exclude those issues from searches, statistics, and/or list custom views?

@edolstra, yes, I would like to join the shepherd team for this PR. And I hope @Profpatsch does too!

ryantm commented:
I made significant edits to this RFC:

1. Removed all references to closing issues
2. Changed stale timeframe to 180 days (needs activity twice a year to not be stale)
3. Added helpful suggestions to the stale marking message

Please take a look again.

zimbatm commented on rfcs/0051-mark-stale-issues.md:
```diff
@@ -0,0 +1,82 @@
+[motivation]: #motivation
+
+The Nixpkgs GitHub page has a large number of open issues causing
+community angst and misrepresenting the responsiveness of the project.
```
Nitpicking a bit, I would prefer if the motivation would be to reduce the number of open issues so that we can focus on the important ones. Something like:

> We have a large number of open issues and PRs that have accumulated over the years. Not all of them are still valid and needs our attention.

ryantm replied:
Thanks; that sounds a lot more positive.

Profpatsch commented on rfcs/0051-mark-stale-issues.md:
```diff
@@ -35,29 +35,43 @@ efforts on the ones that have at least one person interested in them.
+        related code and @ mention them in a comment.
+     2. Ask on the [NixOS Discourse](https://discourse.nixos.org/).
+     3. Ask on the [#nixos channel](irc://irc.freenode.net/#nixos) on
+        [irc.freenode.net](https://freenode.net).
```
Pointing people to the community channels is a great idea!

peti commented:
> We have the following shepherd team: @globin, @grahamc and @peti. @peti, do you want to lead this team?

Yes, I'd be happy to. (Sorry for the delayed response, this fell through the cracks somehow.)

Profpatsch commented:
> And I hope @Profpatsch does too!

ehehe, I was busy over the last days; but I’m sure you three will do a good job of it!

Infinisil commented:
So the only motivation I get for marking issues as stale is that it allows excluding/including them from searches.

However, GitHub already supports that by allowing `updated:<YYYY-MM-DD` in search queries. E.g here are the open issues with no activity in the last 6 months: https://github.com/NixOS/nixpkgs/issues?q=updated%3A%3C2019-03-24+is%3Aopen. In fact, marking issues as stale breaks this search function, because adding a label is seen as an activity to GitHub.

Unless there's an actually useful thing we get from this that GitHub search can't already do, I'm against this.

ryantm commented:
Here are the advantages over the "updated at" search:

1. Automatically see if people are still interested in the issue
2. Automatically say other channels people can use to ask for help on the issue

Infinisil commented:
@ryantm I see, the first point seems like a decent reason, maybe you can add that to the Motivation section?

So question: An issue gets marked as stale with a label and a comment gets posted informing people of this. If somebody is still interested in the issue they can post a comment. Does the stale label get removed again? If so, how will it get removed?

But this kind of seems like it goes the wrong direction, ideally it would first post a comment asking whether the issue is stale, and then add a stale label if this was confirmed.

alyssais commented:
An argument against this that I don’t think has been made yet:

I get too many notifications from Nixpkgs as it is, because of, for example, GitHub’s awful rebase code owner notifying behaviour. I do not want even more irrelevant notifications filling up my inbox because every few days somebody has to make a substanceless comment just to bump one of the many issues I’m subscribed to.

If we do decide to auto close issues (and I hope we don’t), then I propose we at least do the following:

- We never automatically **lock** issues after inactivity. Homebrew does this and it’s extremely demotivating. What am I supposed to do if I have the same issue? Open a dupe?
- It should be possible for anybody to reopen an issue closed due to inactivity. The simplest way to do this would be to have the bot re-open a closed issue if it was commented on, _if and only if_ it was closed by the bot. That way, a human could mark it as closed and it would be closed forever, but it would be easy for anybody to undo false positives by the bot.

In short, actions carried out by the issue closing bot should never be permanent — they should always be overridable by an unprivileged user.

Infinisil commented:
I think we're over the automatically closing issues already and are only considering marking issues as stale now (label + comment)

timokau commented:
> An argument against this that I don’t think has been made yet:
> 
> I get too many notifications from Nixpkgs as it is, because of, for example, GitHub’s awful rebase code owner notifying behaviour. I do not want even more irrelevant notifications filling up my inbox because every few days somebody has to make a substanceless comment just to bump one of the many issues I’m subscribed to.

I would agree if we were actually talking about a matter of days. But most people here seem to be in favor of something more like half a year. In that timeframe I think notifications are reasonable. They may even be a benefit, since everyone interested enough in the issue to subscribe will be reminded that it is still present. That may "nerd-snipe" someone into action or lead someone to make a comment with new information.

7c6f434c commented:
So, what is the optimal order? Bot posts a «triage» comment — timeout — «stale» label — anyone can ask bot to remove the stale label?

By the way, should manual «triage: is this still relevant?» comments count as activity? (Answers to them definitely should)

globin commented:
I personally am fine with the current approach proposed in the RFC, it has slight advantages over doing nothing (https://github.com/NixOS/rfcs/pull/51#issuecomment-534597586), and I can't see this being a negative change. I think any further ideas can be iterated in future RFCs and this should be an improvement for now, especially due to it providing some help to the creator of the issue.

What's your opinion on accepting this @peti, @grahamc and if necessary exploring further ideas on issue organisation in future RFCs?

peti commented:
> I personally am fine with the current approach proposed in the RFC.

Same here. The current proposal has no obvious disadvantages. It does have advantages, though, because it distinguishes issues no-one cares about from those someone still cares about even though the issue has been unresolved for a long time.

Sometimes, I come across an old open issue and my impression is that I *can* close it because it seems obsolete to me. But I usually won't just close it. Instead I'll first leave a comment, asking: "Is this still relevant?" Then I just go and do something else. Which means that if no-one responds to my question, then I'll just forget about the issue and never actually close it. I suppose I could note a TODO reminder somewhere saying "close issue XYZ if no-one has responded by such-and-such date", but I'm not the kind of person who does something like that. 

The "stale" label makes this process easier. When I see that label, I know that the bot has already asked "is this still relevant?" Therefore, if I feel that closing the issue is appropriate, then I can just do it.

> What's your opinion on accepting this @peti, @grahamc and if necessary exploring further ideas on issue organisation in future RFCs?

I am happy to accept the PR as it is now.

JohnAZoidberg commented:
@FRidh 
> However, I think in the RFC process that it is also important to record what we did not agree on. E.g., the original idea was to close issues. We did not agree on doing that. It is, however, a realistic alternative, so I think it can and should be listed there.

Maybe your language is unclear, maybe it is your intended meaning.
I would word it like `we have agreed to not close issues`, which is much stronger than `we have not agreed to close issues`.

Either way, yes, it would be useful to include it in the RFC, alongside a list of our reasons for choosing so.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/fcp-for-rfc-0051-mark-stale-nixpkgs-issues/4600/1

asymmetric commented on rfcs/0051-mark-stale-issues.md:
```diff
@@ -0,0 +1,83 @@
+
+     If this issue is still important to you, we ask that you leave a
+     comment below. Your comment can be as simple as "still important to
+     me". This lets people sorting through the issues see the issues that
```
```suggestion
     me". This lets people sorting through the issues see the ones that
```

ryantm replied:
This part got reworded.

asymmetric commented on rfcs/0051-mark-stale-issues.md:
```diff
@@ -0,0 +1,83 @@
+     comment below. Your comment can be as simple as "still important to
+     me". This lets people sorting through the issues see the issues that
+     at least one person still cares about. Someone will have to do this
+     at most twice a year if there is no other activity.
```
It's not clear what this is referring to. Is someone going to check for stale issues twice a year? If so, should we mention when exactly?

Maybe this whole sentence can actually be removed, as I don't think it's relevant for people who care about an issue.

ryantm replied:
It's referring to the interval the bot would re-mark issues as stale.

alyssais commented:
> Ladies and Gentlemen

Polite reminder that there are people in our community who are neither ladies nor gentlemen. :)

deliciouslytyped commented:
I don't think anyone stated anything of the sort, but I would like to state my opinion that having agreed on marking issues stale, marked-stale issues shouldn't be closed "unnecessarily" either. I agree with Profpatsch's statement on the nature of the issue tracker - though at the time he was commenting on auto-closing.

I'm left a bit uneasy by the noise the bot will be adding, however the half-year interval might make that tolerable.

I'm not sure the reasoning is sound, but per @Zimmi48 , 25% of issues are not closed within 500 days - nixpkgs currently has 5000 open pull requests, are 1250 going to retroactively get "spammed" with stale markings?

ryantm commented on rfcs/0051-mark-stale-issues.md:
```diff
@@ -0,0 +1,83 @@
+   markComment: >
+     Thank you for your contributions.
+
+     This has been automatically marked as stale because it has had no
```
Thanks; I resolved this by removing "issue" from the stale message.

Profpatsch commented:
> I'm left a bit uneasy by the noise the bot will be adding, however the half-year interval might make that tolerable. Edit: We're operating at scale here, so small proportional operations end up large in the absolute. For example, a lot of core contributors probably have comments in a large number of stale deep issues.

For reference, I did this by hand a few times. Just ordered issues by “least recently updated” and started pinging people, a hundred issues at a time, something like “(triage) is this still relevant”.

When I saw an issue that had been triaged before and received no updates afterwards, I’d usually close them (but only depending on the issue, using my good (human) judgement.

The bot would do the first step automatically, which is a good thing in my book.

domenkozar commented:
FCP ended, RFC is accepted.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/suggestion-to-help-maintainers-with-nixpkgs-prs-bloat/4794/12

abathur commented:
(I realize this is settled. Not trying to re-litigate, but want to comment where it might be seen for future related efforts.)

I agree that auto-dismissing issues just for staleness is frustrating. I think the right call was made here.

That said, elsewhere I implemented a way to:
- Manually mark issues that aren't actionable without reporter feedback (when requesting that feedback during triage).
- Yank them from the "triage queue" (save N maintainers from wasting tiny slices of time/effort on unactionable issues) until the reporter responds.
- Auto-dismiss them if the reporter doesn't.

It's just another small discretionary tool to focus time/effort.

Zimmi48 commented:
> elsewhere I implemented

@abathur Would you mind providing a link if this tool is available online? It could be useful to other projects, beyond just nixpkgs.

abathur commented:
> > elsewhere I implemented
> 
> @abathur Would you mind providing a link if this tool is available online? It could be useful to other projects, beyond just nixpkgs.

It isn't, sorry (though I agree)--it's a component of the idiomatic text-based issue tracker for a MUD I work on.

doronbehar commented:
Hey all, was the bot deployed lately? I noticed it popping in a few issues, but I never saw it before.

I'm looking for a TL;DR: if there's a PR still relevant and it has been marked as stale by the bot, is there anything the author of it can do?

lheckemann commented:
@doronbehar as the bot mentions in its messages, commenting will "unstale" issues and PRs.

joepie91 commented:
I'd completely missed this RFC somehow. It's good to see that closing issues was removed, but I still see two issues with this:

1. It's not clear at all from the bot message that it *doesn't* auto-close. What does "marked as stale" mean exactly? This is unclear, and considering the usual application of stale-bots, I suspect that I won't be the only one who initially assumed that it meant an impending auto-close.
2. "Automatically see if people are still interested in the issue" sounds great on paper, but what it means in practice is that someone whose issue remains unfixed for a long time is constantly asked to keep putting work into an issue that's not getting addressed on the nixpkgs side at all. That comes across as rather user-hostile, even if it happens once every half year.

abathur commented:
@joepie91 I guess 1 can be fixed by clarifying the message--do you see 2 as intrinsic, or do you think it could also be fixed with (more polite/personable?) phrasing? 

Here's one for reference:


> Thank you for your contributions.
> This has been automatically marked as stale because it has had no activity for 180 days.
> If this is still important to you, we ask that you leave a comment below. Your comment can be as simple as "still important to me". This lets people see that at least one person still cares about this. Someone will have to do this at most twice a year if there is no other activity.
> Here are suggestions that might help resolve this more quickly:
> 
> 1. Search for maintainers and people that previously touched the
> related code and @ mention them in a comment.
> 2. Ask on the NixOS Discourse. 3. Ask on the #nixos channel on irc.freenode.net.


Here's a rephrase--curious if you think it moves the needle: 

> To help our human contributors focus on the most-relevant reports, I check up on issues that have been quiet for 6 months to see if they're still relevant. I added the `2. status: stale` label to this issue for now, but I'll remove the label if anyone comments to let me know it's still relevant.
> 
> I'm not allowed to close issues, so I could use your help closing this if you know it's no longer needed.
> 
> If this issue isn't getting enough attention, there's nothing wrong with asking for a little help:
> 1. Search for maintainers and people that previously touched the related code and @ mention them in a comment.
> 2. Ask on the NixOS Discourse. 
> 3. Ask on the #nixos channel on irc.freenode.net.

I guess there may be other small improvements:
- A progressive +3mo backoff could be nice (but the bot config doesn't look like the stale bot has support).
- Expand the list of exempt labels so that issues in stasis for known reasons don't get poked pointlessly.
- Similar to above, it might be interesting (though I suspect also not supported) to have separate bots/rules/messages for different labels. I could imagine tailoring the message and interval for things blocked by a PR/issue, waiting for upstream, WIP, merge conflicts, etc.

doronbehar commented:
@abathur I like the 2nd phrasing a bit more and I tend to agree with @joepie91 . In general though, I don't see a point in a stale bot when Nixpkgs has so many stale PRs that are _worth_ looking at. The main reason they become stale is not because their author doesn't care anymore, but because their author doesn't feel comfortable to either:

- Ping a committer because their PR got an approval from a reviewer without merge access.
- Ping a reviewer that started the review but stopped in midway.
- Ping a committer once more.

Reading these messages from the bot in my INBOX makes _me_: 

> (sigh) no committer has reached that PR yet... and probably no one will because it's buried down the constant feed of newer PRs....

vs:

> Oh! Thanks there's this bot now because I forgot about this PR and now I know what to do!".

Moreover, who says there _has_ to be a message? As long as I'm not a committer, I don't see a reason not to apply a filter to my INBOX that will delete these messages to spare me the disappointment.

I wish this bot was a little bit more sophisticated then that. I once tried to suggest such an idea at https://discourse.nixos.org/t/suggestion-to-help-maintainers-with-nixpkgs-prs-bloat/4794 but the discussion got mixed in different topics.

TL;DL:

Autolabel PRs whose author _responds_ to reviews.

And when I think about it once more. Maybe also:

Make the _bot_ ping committers every once in a while, because it feels a bit [offensive](https://github.com/NixOS/nixpkgs/pull/85922#issuecomment-621220386) to do it yourself.

With such a bot, committers would be able to use a super simple filter to the list that will consist of nothing more then a `label:author last responded`. This, along with a `sort:older` is a perfect setup for reviewing PRs _worth looking at_.

vcunat commented:
I'm personally saddened that in the last 12h my notifications get way more activity from this bot than from any humans, but hopefully that's a short-term state...

ryantm commented:
> I'm personally saddened that in the last 12h my notifications get way more activity from this bot than from any humans, but hopefully that's a short-term state...

I expect this is consequence of how it was misconfigured and then fixed yesterday. Sorry about that.

joepie91 commented:
> Here's a rephrase--curious if you think it moves the needle:

This would be an improvement. I think the "not allowed to" phrasing is a bit weird, though - "I won't close issues" seems like a clearer and less awkward way to state that.

However...

> do you see 2 as intrinsic, or do you think it could also be fixed with (more polite/personable?) phrasing?

Yes, I do think it's an intrinsic problem. And while I hadn't really considered it from this angle before, I actually really like @doronbehar's suggestion; labeling things positively rather than negatively.

Aside from generally making prioritization easier as mentioned, it also sends a much better message to users - instead of the subtext of "hey we're kinda annoyed by open issues/PRs" that a stale bot has, the subtext would be "hey we see that there's been some progress here, let me get someone with the right access for you!".

I think that fits much better into the general goal of keeping the NixOS community a friendly place, and is much less likely to frustrate users. It would also drive the right incentives; encouraging users to remain actionably responsive, rather than encouraging them to stay quiet if they encounter any issues in the future because their issue will die in `2.status: stale` anyway.

timokau commented:
@doronbehar writing such a bot has been on my todo-list for a long time. Unfortunately my todo list has a serious task starvation problem...

As I said in https://github.com/NixOS/rfcs/pull/30#issuecomment-615194338, if you or anyone wants to discuss this feel free to reach out. Bonus points if you have some free time available and want to work on this. You don't need to be a "core member", you don't need special permissions. On the contrary, having less permissions gives you a better perspective in this case.

/OT

8573 commented:
Three minor, disconnected observations from my very minor contributor's perspective:

1. I join in supporting @abathur's phrasing.

2. What it says aside, I've found the bot to have its use as an automated version of @Profpatsch going around asking "(triage) Is this still a problem": it's reminded me to check whether I still have issues, and in at least one case I didn't (https://github.com/NixOS/nixpkgs/issues/47385).

3. When not looking carefully, I've confused it with @timokau... :p

asymmetric commented:
Could someone remove the `status: FCP` label from this PR?

timokau commented:
>     3. When not looking carefully, I've confused it with @timokau... :p

Why is that? I might be missing something here.

8573 commented:
> Why is that? I might be missing something here.

Your avatar and its both have rectangular heads, chest-mounted monitors, claw-hands, and ball-tipped antennæ atop the head. :)

bhipple commented:
This bot is great; I found and closed several long-since resolved issues from its pings!  

The above wording does seem like an improvement, though.

nh2 commented:
A little less enthusiastic feedback:

The bot does create quite a bit of work for me. Just in the last 4 days I've received 60 emails:

![image](https://user-images.githubusercontent.com/399535/83827684-31b3b480-a6df-11ea-9b68-44ec40d69377.png)

On pretty much all of them I have to write `still important to me`, because bugs rarely fix themselves (I'm glad though that @bhipple found some; I dont seem to be that lucky).

I find it somewhat stressful to go through that amount of (mainly unnecessary) work, because I _do_ want to point out when issues are not stale, and remove the label to make that clear.

As an aside, for some reason `NixOS/nixpkgs` does not appear in https://github.com/notifications/subscriptions for me, so I cannot really tell how many nixpkgs issues I am subscribed to in total (to estimate total workload):

![image](https://user-images.githubusercontent.com/399535/83830037-83ab0900-a6e4-11ea-92b4-9fc2c44198fa.png)

It feels like the task of marking issues that are still important to me as non-stale is something I soon have to automate, because it takes a significant amount of my daily time. :thinking:

Zimmi48 commented:
In my case, the bot helped me recover an issue (NixOS/nixpkgs#24613) that I had commented in 2017 to say: it can be closed. The only other comment it had received since then was someone else saying: it can be closed. It's a bit sad that nobody with the right permission had reacted to close it but now I have triage permissions, so I was finally able to close it!

:+1: for modifying the message to be more friendly and explicit on what people could do

domenkozar commented:
I agree it's annoying (I get multiple dozen emails per day), but once the bot goes through the backlog the noise will turn down.

vcunat commented:
I guess this has to affect most of contributors that have been very active over long term.

@nh2: perhaps you mistakenly thought that the bot would close the issues?  If some are very important to you it surely does make sense to express that on the issue, but if not... keeping the "stale" label doesn't seem a big deal to me.  (arguably, just the fact that noone is working on the issue for many months)

> As an aside, for some reason NixOS/nixpkgs does not appear in https://github.com/notifications/subscriptions for me, so I cannot really tell how many nixpkgs issues I am subscribed to in total (to estimate total workload):

It appears to me, but "1000+" isn't very informative.  You can try https://github.com/NixOS/nixpkgs/issues?q=involves%3Anh2

Profpatsch commented:
> What it says aside, I've found the bot to have its use as an automated
version of @Profpatsch <https://github.com/Profpatsch> going around asking
"(triage) Is this still a problem": it's reminded me to check whether I
still have issues, and in at least one case I didn't (NixOS/nixpkgs#47385
<https://github.com/NixOS/nixpkgs/issues/47385>).

Yes, that’s why I’m glad this exists now. Regular triage was purely
mechanical (and I didn’t even do it regularly, more like once a year).

Improving the phrasing is always possible, thus I’m glad people are
discussing it.
I agree that it should be formulated as positively as possible. in general,
you don’t want to give people negative emotions if not necessary, and
positive and encouraging phrasing contributes a lot to that.


On Thu, Jun 4, 2020 at 3:59 AM 8573 <notifications@github.com> wrote:

> Three minor, disconnected observations from my very minor contributor's
> perspective:
>
>    1.
>
>    I join in supporting @abathur <https://github.com/abathur>'s phrasing.
>    2.
>
>    What it says aside, I've found the bot to have its use as an automated
>    version of @Profpatsch <https://github.com/Profpatsch> going around
>    asking "(triage) Is this still a problem": it's reminded me to check
>    whether I still have issues, and in at least one case I didn't (
>    NixOS/nixpkgs#47385 <https://github.com/NixOS/nixpkgs/issues/47385>).
>    3.
>
>    When not looking carefully, I've confused it with @timokau
>    <https://github.com/timokau>... :p
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/rfcs/pull/51#issuecomment-638555446>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAYB5ZVQI6QWWEYIFD6UINTRU35YTANCNFSM4IPHMSPQ>
> .
>

nh2 commented:
> I agree it's annoying (I get multiple dozen emails per day), but once the bot goes through the backlog the noise will turn down.

I hope so :) Can anybody clarify the working mechanism of the bot:

Via which method does it decide how to comment on issues that have no activity for _longer_ than 6 months? E.g. [this one](https://github.com/NixOS/nixpkgs/issues/22799#issuecomment-326570339) had the last human comment in 2017. Is it slowly working through a backlog? If yes, is there a way to see how far it is, or know when it's done?

The fact that old issues were not all commented on instantly but over a time of multiple days may falsely have created the impression that what I see in my inbox now is the "stable throughput" of just the "every 180 days" logic.

If it's really just an initial burst and afterwards the daily volume is much lower, that's probably fine!

> @nh2: perhaps you mistakenly thought that the bot would close the issues? If some are very important to you it surely does make sense to express that on the issue, but if not... keeping the "stale" label doesn't seem a big deal to me. (arguably, just the fact that noone is working on the issue for many months)

I did not think that it would close the issues; if I remember correctly, I was part of a discussion before where most people agreed that a stale-bot should only set labels, not auto-close, and I was also an advocate of labeling-only.

I do feel obliged to comment on these issues as the bot asks, not because they are "very important" to me but because it would be a waste of time (mine and others') if somebody filed the same bug twice or did the same investigation twice, and to make clear to others that yes, this bug still exists (in other words, keeping issue tracker information accurate).

>  You can try https://github.com/NixOS/nixpkgs/issues?q=involves%3Anh2

Unfortunately that method [does not work](
https://stackoverflow.com/questions/25438721/how-can-i-see-all-the-issues-im-watching-on-github/33478315#33478315) for issues where you've clicked the **Subscribe** button on the right, only for issues where you're `author`, `assignee`, `mentions` or `commenter`. So you'd get an underapproximation.

I checked yesterday; Github's API does not seem to provide a way to get at the total number of subscribed issues. I tried `curl 'https://api.github.com/issues?filter=subscribed'` but it only shows a few per page, and pagination there works only up to `page=24`, above that it just returns `[]`.

That said, the `involves` filter shows a 152 open issues for me, so if I assume roughly the same amount of non-`involves` subscriptions, that would pan out at ~2 notifications per day, which suggests that @domenkozar is right and that what we observe is just a backlog.

abathur commented:
> I hope so :) Can anybody clarify the working mechanism of the bot:
> 
> Via which method does it decide how to comment on issues that have no activity for _longer_ than 6 months? E.g. [this one](https://github.com/NixOS/nixpkgs/issues/22799#issuecomment-326570339) had the last human comment in 2017. Is it slowly working through a backlog? If yes, is there a way to see how far it is, or know when it's done?

With the asterisk that I'm lazily looking at the issues (i.e., [issues with stale label by most-recently-updated](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%222.status%3A+stale%22+sort%3Aupdated-desc)) to infer what it's doing:
- it looks like it's just working from least-recently-commented to most-recently commented
- as of the last round (about an hour ago, so I think it's about to run a new batch) it's up to July 5, 2017

I'll update this in a few minutes with a little more context/meta.

timokau commented:
At least the PR part can (and hopefully will) be improved by [marvin](https://discourse.nixos.org/t/rfc-marvin-mk2-the-pull-request-bot/7547). Instead of marking PRs as stale, we'd assign labels to it which indicate for whom it is actionable (author, reviewer, merger). Then stale labels would either be unnecessary or could be restricted to `needs_work` PRs.

I hope this will already help in getting you a reviewer for your PRs, since reviewers can now easily find actionable PRs and everybody can help out, regardless of whether they have the commit bit or not.

Once that is working, we could explore https://github.com/timokau/marvin-mk2/issues/2 to get some steady "baseline-throughput" of PRs. We should make sure that people don't sign up for too much there though to avoid burnout.

doronbehar commented:
I once had an idea, which may be worth mentioning here, also considering https://github.com/timokau/marvin-mk2/issues/2 :

What if there was a website or a database that I'd upload to it the list of packages in my `system.environmentPackages` and the system services I enable. Then, a bot would ping me (perhaps privately) whenever there's a PR with some kind of update / change to a package I use. This will invite me to test the changes and if I'll respond then, my feedback should be considered valuable because I'm an actual user of the package service being changed.

ryantm commented:
@doronbehar I'm all for improving the stale bot wording! I think we can discuss the details in a regular nixpkgs PR.

doronbehar commented:
OK, here's my suggestion: https://github.com/NixOS/nixpkgs/pull/92254

Ericson2314 commented:
https://github.com/NixOS/rfcs/pull/124 Follow-up RFC for Nix itself is here.
